shared/middleware/table_coerce.rb

```diff
@@ -0,0 +1,105 @@
+# A set of methods for coercing data in our table to specific types. In general,
+# the goal when converting is to change values where we can, but leave alone
+# those where we don't know how to convert (i.e. I can't convert "123" to a
+# boolean.
+
+module TableCoerce
+  # Given a set of records, and a list of columns, attempt to convert each
+  # column to a particular type. If we can convert the entire column to
+  # number or boolean, we'll do that, otherwise we'll convert all values to strings.
+  def TableCoerce.coerce(records, columns)
+    # first do a pass to determine types. we can't do coercion right away since
+    # we dont want to change anything if we cant coerce the entire column
+    # begin with no type definitions
+    column_types = TableCoerce.column_types(records, columns)
+
+    # second pass to convert values. by default, values are all strings
+    records.each do |row|
+      columns.each_index do |index|
+        type = column_types[index]
+        col = columns[index]
+        val = row[col]
+        if type == :boolean
+          row[col] = TableCoerce.to_boolean(val)
+        elsif type == :number
+          row[col] = TableCoerce.to_number(val)
+        end
+      end
+    end
+
+    records
+  end
+
+  def TableCoerce.column_types(records, columns)
+    # do an initial pass to determine column types
+    return columns.map do |col|
+      if records.empty?
+        :string
+      elsif !records.any? { |record| !TableCoerce.is_boolean?(record[col]) }
+        :boolean
+      elsif !records.any? { |record| !TableCoerce.is_number?(record[col]) }
+        :number
+      else
+        :string
+      end
+    end
+  end
+
+  # Attempts to coerce a single column to the given type.
+  # @return {[Array, boolean]} The new set of records, and a boolean indicating
+  #  whether we were able to convert every single row.
+  def TableCoerce.coerce_column(records, column_name, type)
+    all_converted = true
+    records.map do |record|
+      val = record[column_name]
+      if type == :boolean
+        convertable = TableCoerce.is_boolean?(val)
+        val = TableCoerce.to_boolean(val) if convertable
+        all_converted = false unless convertable
+      elsif type == :number
+        convertable = TableCoerce.is_number?(val)
+        val = TableCoerce.to_number(val) if convertable
+        all_converted = false unless convertable
+      elsif type == :string
+        val = val.to_s
+      end
+      record[column_name] = val
+      record
+    end
+    [records, all_converted]
+  end
+
+  # @param val [String|Boolean|Number]
+  def TableCoerce.is_boolean?(val)
+    return true if val == true || val == false
+    return false unless val.is_a?(String)
+    return !/true/i.match(val).nil? || !/false/i.match(val).nil?
+  end
```

Comment:
Quite right. I'm not sure why I wasn't doing that from the beginning. Prob a good thing for me to add a test for too :)